### PR TITLE
Improve get_psf_kernel performance

### PR DESCRIPTION
**Pull request by QRemy**
Improve get_psf_kernel performance, split of #5297

### CODE REVIEW

File: gammapy/irf/psf/map.py [hunk 1]
```diff
@@ -24,17 +24,19 @@ def _psf_upsampling_factor(psf, geom, position, energy=None, precision_factor=12
...
-    if isinstance(geom, HpxGeom):
-        factor = int(2 ** np.ceil(np.log(factor) / np.log(2)))
-    return factor
+    factors = []
+    for psf_r68 in psf_r68s:
+        base_factor = (2 * psf_r68 / geom.pixel_scales.max()).to_value("")
+        factor = np.minimum(
+            int(np.ceil(precision_factor / base_factor)), PSF_MAX_OVERSAMPLING
+        )
+        if isinstance(geom, HpxGeom):
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/irf/psf/map.py:36]: No as the computation for the `HpxGeom ` use the previously computed factor and round it up to next power of 2.

[the remark this answers is not in the document]
Comment (VascoSch92) [reply gammapy/irf/psf/map.py:36]: ah ok i see

File: gammapy/irf/psf/map.py [hunk 2]
```diff
@@ -283,43 +286,64 @@ def get_psf_kernel(
...
-        coords = geom.get_coord(sparse=True)
-        rad = coords.skycoord.separation(geom.center_skydir)
+        nr = len(radii)
+        if factor is None:  # TODO: this remove and the else once factor is deprecated
+            factor = _psf_upsampling_factor(self, geom, position, precision_factor)
+        else:
+            factor = [factor] * nr
+        geom = geom.to_odd_npix(max_radius=max_radius)
+        kernel_map = Map.from_geom(geom=geom)
+        for im, ind in zip(kernel_map.iter_by_image(keepdims=True), range(nr)):
```
[the remark this answers is not in the document]
Comment (registerrier) [reply gammapy/irf/psf/map.py:321]: or even 
`for im, fact, radius in zip(kernel_map.iter_by_image(keepdims=True), factor, radii)`

File: gammapy/irf/psf/map.py [hunk 3]
```diff
@@ -270,10 +272,11 @@ def get_psf_kernel(
...
             Maximum angular size of the kernel map.
+            Default is None and it will be computed for the `containment` fraction set.
         containment : float, optional
-            Containment fraction to use as size of the kernel. The max. radius
-            across all energies is used. The radius can be overwritten using
-            the `max_radius` argument. Default is 0.999.
+            Containment fraction to use as size of the kernel.
+            The radius can be overwritten using the `max_radius` argument.
+            Default is 0.999.
         factor : int, optional
```
Comment (registerrier) [inline gammapy/irf/psf/map.py:280]: If you plan to deprecate `factor` you should add the proper decorator.

**Review by registerrier — DISMISSED** [review]
Thanks @QRemy. Looks good to me. I have left a few minor comments.

File: gammapy/irf/psf/map.py (hunk 3 above)
Comment (QRemy) [reply gammapy/irf/psf/map.py:280]: it is already deprecated since v1.2

**Pull request merged**